README.md
# Kaggle-Autonomous_Driving

## How to run this project:
1. Set up pytorch environment.
2. run ```python /Kaggle-Autonomous_Driving/Project/task/Autonomous_Driving.py```

ChangeLog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.8] -2019-12-28

### Added

- Add focal loss on mask

### Fixed

- Heatmap now works well

## [0.0.7] -2019-12-17

### Added

- Data Agumentation including blur, noise and flip
- Heatmap of mask
- Now we can decide if to use optimization of prediction (shorten time)
- Another base model: wide-resnet-101
- Use mask dataset information

### Changed

- Delete additional side of input image
- Increase threshold of sigmoid output
- Use input size of 512, batch size of 8

### Fixed
- Flip function works well now

## [0.0.6] -2019-11-28

### Added

- Stolen figure saving method from Kailiang.
- Normalize image in DataProcessing module.
- Assign weight on regr loss and mask loss respectively.

### Changed

- Save figures instead of showing them.
- Add regularization in model.

### Fixed

- Make resnet really learnable.

## [0.0.5] -2019-11-27

### Fixed

- Add system path. Now the project can be ran in terminal.


## [0.0.4] - 2019-11-26

### Changed

- Make upsample layer learnable.
- Add visulization function. 
- Fix some minor bugs.
- Improve structure of project.

### Results

- Submit ```resnetx101-Unet``` without learnable upsampling layer, get score of ```0.077```. :hankey:
- Submit ```resnetx101-Unet``` with learnable upsampling layer, get score of ```0.103```. :blush:

## [0.0.3] - 2019-11-24

### Changed 

- Change efficient-net to resnetx101.


## [0.0.2] - 2019-11-22

### Added

- Reconstruct [CenterNet_Baseline](https://www.kaggle.com/hocop1/centernet-baseline) to be the baseline of our project.

## [0.0.1] - 2019-11-20

### Added

- Run notebook stolen from [CenterNet_Baseline](https://www.kaggle.com/hocop1/centernet-baseline).

Project/src/Routine.py
import numpy as np  # linear algebra
import pandas as pd  # data processing, CSV file I/O (e.g. pd.read_csv)
from tqdm import tqdm  #_notebook as tqdm
import matplotlib.pyplot as plt
import gc
from sklearn.model_selection import train_test_split
import torch
import torch.optim as optim
from torch.optim import lr_scheduler
from torch.utils.data import DataLoader
from src.DataProcessing import DataProcessing
from termcolor import colored
from math import sqrt, acos, pi, sin, cos
from scipy.spatial.transform import Rotation as R
from sklearn.metrics import average_precision_score
import os
from datetime import datetime
import cv2


class Routine(object):
    def __init__(self, settings, model):
        self._settings = settings
        self.data_loader = DataProcessing
        self.train_data = None
        self.path = self._settings.path
        self.device = self._settings.device
        self.Model = model
        self.xzy_slope = None

    def run(self):
        dir_name = self.make_log_dir()
        df_train, df_dev, df_test, train_loader, dev_loader, test_loader, train_dataset, dev_dataset, test_dataset = self.data_loading()
        self.show_image(train_dataset[0][0])
        self.show_image(train_dataset[0][1])
        self.show_image(train_dataset[0][2])

        # for idx in range(2):
        #     fig, axes = plt.subplots(1, 2, figsize=(20, 20))
        #     for ax_i in range(2):
        #         img, mask, regr = dev_dataset[idx]
        #         coords = dev_dataset.extract_coords(np.concatenate([mask[None], regr], 0), ax_i == 1) #  mask[None] == np.newaxis
        #         train_images_dir = self.path + 'train_images/{}.jpg'
        #         img = dev_dataset.imread(train_images_dir.format(df_dev['ImageId'].iloc[idx]))
        #         if ax_i == 1:
        #             img = img[:, ::-1]
        #         axes[ax_i].set_title('Flip = {}'.format(ax_i == 1))
        #         axes[ax_i].imshow(dev_dataset.visualize(img, coords))
        #     plt.show()

        model = self.Model(8, self._settings).to(self.device)
        optimizer = optim.Adam(model.parameters(), lr=self._settings.lr, weight_decay=self._settings.reg_factor)
        exp_lr_scheduler = lr_scheduler.StepLR(optimizer, step_size=self._settings.lr_decay_epoch * len(train_loader), gamma=0.1)

        history = pd.DataFrame()
        for epoch in range(self._settings.epoch):
            self.train_model(optimizer, model, exp_lr_scheduler, epoch, train_loader, history)
            self.evaluate_model(model, epoch, dev_loader, history)
        if self._settings.save_model:
            torch.save(model.state_dict(), dir_name + '/model.pth')
        history['train_loss'].iloc[100:].plot().get_figure().savefig(dir_name + '/train_loss.png')
        plt.cla()
        series = history.dropna()['dev_loss']
        plt.scatter(series.index, series)
        plt.savefig(dir_name + '/dev_loss.png')
        torch.cuda.empty_cache()
        gc.collect()

        dev_dataset.train_dataset = self.train_data
        dev_dataset.point_regr()
        for idx in range(5):
           img, mask, regr = dev_dataset[idx]
           output = model(torch.tensor(img[None]).to(self._settings.device)).data.cpu().numpy()
           coords_pred = dev_dataset.extract_coords(prediction=output[0])
           coords_true = dev_dataset.extract_coords(prediction=np.concatenate([mask[None], regr], 0))
           dev_dataset.show_result(idx, coords_true, coords_pred, df_dev, dir_name)

        predictions = []
        model.eval()

        #        Run Test        #
        for img, _, _ in tqdm(test_loader):
            with torch.no_grad():
                output = model(img.to(self.device))
            output = output.data.cpu().numpy()
            for out in output:
                coords = dev_dataset.extract_coords(out)
                s = self.data_loader.coords2str(coords)
                predictions.append(s)

        test = pd.read_csv(self.path + 'sample_submission.csv', nrows=getattr(self._settings, 'nrof_test_data', None))
        test['PredictionString'] = predictions
        test.to_csv(dir_name + '/predictions.csv', index=False)
        test.head()
        # todo: fix bug in local metric calculation function
        # self.local_metric(test)
        plt.show()

    def data_loading(self):
        train = pd.read_csv(self.path + 'train.csv')
        print(colored('Original Train Data Shape:', 'green'), train.shape)
        test = pd.read_csv(self.path + 'sample_submission.csv', nrows=getattr(self._settings, 'nrof_test_data', None))
        print(colored('Original Test Train Data Shape:', 'green'), test.shape)
        camera_matrix = np.array([[2304.5479, 0, 1686.2379],
                                       [0, 2305.8757, 1354.9849],
                                       [0, 0, 1]], dtype=np.float32)
        camera_matrix_inv = np.linalg.inv(camera_matrix)
        train_data = train.sample(frac=self._settings.data_sample_factor, random_state=1, axis=0)
        df_test = test
        self.train_data = train_data
        train_images_dir = self.path + 'train_images/{}.jpg'
        test_images_dir = self.path + 'test_images/{}.jpg'
        # Create dataset objects
        df_train, df_dev = train_test_split(train_data, test_size=self._settings.validation_data_ratio, random_state=6)
        print(colored('Train Data Shape:', 'green'), df_train.shape)
        print(colored('Evaluation Data Shape:', 'green'), df_dev.shape)
        print(colored('Test Data Shape:', 'green'), df_test.shape)
        train_dataset = self.data_loader(self._settings, df_train, train_images_dir, data_agument=self._settings.data_agument, training=True)
        dev_dataset = self.data_loader(self._settings, df_dev, train_images_dir, training=False)
        test_dataset = self.data_loader(self._settings, df_test, test_images_dir, training=False)

        train_loader = DataLoader(dataset=train_dataset, batch_size=self._settings.batch_size, shuffle=True,
                                  num_workers=0)
        dev_loader = DataLoader(dataset=dev_dataset, batch_size=self._settings.batch_size, shuffle=False, num_workers=0)
        test_loader = DataLoader(dataset=test_dataset, batch_size=4, shuffle=False, num_workers=0)
        return df_train, df_dev, df_test, train_loader, dev_loader, test_loader, train_dataset, dev_dataset, test_dataset

    def criterion(self, prediction, mask, regr, weight=0.5, size_average=True):
        # Binary mask loss
        pred_mask = torch.sigmoid(prediction[:, 0])
        # mask_loss = mask * torch.log(pred_mask + 1e-12) + (1 - mask) * torch.log(1 - pred_mask + 1e-12)

        # Focal loss
        gamma = 0.5
        alpha = 0.5
        mask_loss = mask * torch.log(pred_mask + 1e-12) * alpha * (1 - pred_mask) ** gamma + \
                    (1 - mask) * torch.log(1 - pred_mask + 1e-12) * (1 - alpha) * pred_mask ** gamma
        mask_loss = -mask_loss.mean(0).sum()
        # Regression L1 loss
        pred_regr = prediction[:, 1:]
        regr_loss = (torch.abs(pred_regr - regr).sum(1) * mask).sum(1).sum(1) / mask.sum(1).sum(1)
        regr_loss = regr_loss.mean(0)
        # Sum
        loss = weight * mask_loss + (1 - weight) * regr_loss
        if not size_average:
            loss *= prediction.shape[0]
        return loss

    def train_model(self, optimizer, model, exp_lr_scheduler, epoch, train_loader, history=None):
        model.train()
        total_loss = 0
        for batch_idx, (img_batch, mask_batch, regr_batch) in enumerate(tqdm(train_loader)):
            img_batch = img_batch.to(self.device)
            mask_batch = mask_batch.to(self.device)
            regr_batch = regr_batch.to(self.device)
            optimizer.zero_grad()
            output = model(img_batch)
            loss = self.criterion(output, mask_batch, regr_batch)
            total_loss += loss.data
            if history is not None:
                history.loc[epoch + batch_idx / len(train_loader), 'train_loss'] = loss.data.cpu().numpy()
            loss.backward()
            optimizer.step()
            exp_lr_scheduler.step()
        print('Train Epoch: {} \tLR: {:.6f}\tLoss: {:.6f}'.format(epoch, optimizer.state_dict()['param_groups'][0]['lr'], total_loss/len(train_loader.dataset)))

    def evaluate_model(self, model, epoch, dev_loader, history=None):
        model.eval()
        loss = 0

        with torch.no_grad():
            for img_batch, mask_batch, regr_batch in dev_loader:
                img_batch = img_batch.to(self.device)
                mask_batch = mask_batch.to(self.device)
                regr_batch = regr_batch.to(self.device)

                output = model(img_batch)

                loss += self.criterion(output, mask_batch, regr_batch, size_average=False).data

        loss /= len(dev_loader.dataset)

        if history is not None:
            history.loc[epoch, 'dev_loss'] = loss.cpu().numpy()

        print('Dev loss: {:.4f}'.format(loss))

    def local_metric(self, test_prediction):

        thres_tr_list = [0.1, 0.09, 0.08, 0.07, 0.06, 0.05, 0.04, 0.03, 0.02, 0.01]
        thres_ro_list = [50, 45, 40, 35, 30, 25, 20, 15, 10, 5]

        def TranslationDistance(p, g, abs_dist=False):
            dx = p['x'] - g['x']
            dy = p['y'] - g['y']
            dz = p['z'] - g['z']
            diff0 = (g['x'] ** 2 + g['y'] ** 2 + g['z'] ** 2) ** 0.5
            diff1 = (dx ** 2 + dy ** 2 + dz ** 2) ** 0.5
            if abs_dist:
                diff = diff1
            else:
                diff = diff1 / diff0
            return diff

        def RotationDistance(p, g):
            true = [g['pitch'], g['yaw'], g['roll']]
            pred = [p['pitch'], p['yaw'], p['roll']]
            q1 = R.from_euler('xyz', true)
            q2 = R.from_euler('xyz', pred)
            diff = R.inv(q2) * q1
            W = np.clip(diff.as_quat()[-1], -1., 1.)

            # in the official metrics code:
            # https://www.kaggle.com/c/pku-autonomous-driving/overview/evaluation
            #   return Object3D.RadianToDegree( Math.Acos(diff.W) )
            # this code treat θ and θ+2π differntly.
            # So this should be fixed as follows.
            W = (acos(W) * 360) / pi
            if W > 180:
                W = 360 - W
            return W

        def expand_df(df, PredictionStringCols):
            df = df.dropna().copy()
            df['NumCars'] = [int((x.count(' ') + 1) / 7) for x in df['PredictionString']]

            image_id_expanded = [item for item, count in zip(df['ImageId'], df['NumCars']) for i in range(count)]
            prediction_strings_expanded = df['PredictionString'].str.split(' ', expand=True).values.reshape(-1,
                                                                                                            7).astype(
                float)
            prediction_strings_expanded = prediction_strings_expanded[
                ~np.isnan(prediction_strings_expanded).all(axis=1)]
            df = pd.DataFrame(
                {
                    'ImageId': image_id_expanded,
                    PredictionStringCols[0]: prediction_strings_expanded[:, 0],
                    PredictionStringCols[1]: prediction_strings_expanded[:, 1],
                    PredictionStringCols[2]: prediction_strings_expanded[:, 2],
                    PredictionStringCols[3]: prediction_strings_expanded[:, 3],
                    PredictionStringCols[4]: prediction_strings_expanded[:, 4],
                    PredictionStringCols[5]: prediction_strings_expanded[:, 5],
                    PredictionStringCols[6]: prediction_strings_expanded[:, 6]
                })
            return df

        def check_match(idx):
            keep_gt = False
            thre_tr_dist = thres_tr_list[idx]
            thre_ro_dist = thres_ro_list[idx]
            train_dict = {imgID: self.data_loader.str2coords(s, names=['carid_or_score', 'pitch', 'yaw', 'roll', 'x', 'y', 'z']) for
                          imgID, s in zip(train_df['ImageId'], train_df['PredictionString'])}
            valid_dict = {imgID: self.data_loader.str2coords(s, names=['pitch', 'yaw', 'roll', 'x', 'y', 'z', 'carid_or_score']) for
                          imgID, s in zip(valid_df['ImageId'], valid_df['PredictionString'])}
            result_flg = []  # 1 for TP, 0 for FP
            scores = []
            MAX_VAL = 10 ** 10
            for img_id in valid_dict:
                for pcar in sorted(valid_dict[img_id], key=lambda x: -x['carid_or_score']):
                    # find nearest GT
                    min_tr_dist = MAX_VAL
                    min_idx = -1
                    for idx, gcar in enumerate(train_dict[img_id]):
                        tr_dist = TranslationDistance(pcar, gcar)
                        if tr_dist < min_tr_dist:
                            min_tr_dist = tr_dist
                            min_ro_dist = RotationDistance(pcar, gcar)
                            min_idx = idx

                    # set the result
                    if min_tr_dist < thre_tr_dist and min_ro_dist < thre_ro_dist:
                        if not keep_gt:
                            train_dict[img_id].pop(min_idx)
                        result_flg.append(1)
                    else:
                        result_flg.append(0)
                    scores.append(pcar['carid_or_score'])

            return result_flg, scores

        # expanded_valid_df = expand_df(test_prediction, ['pitch', 'yaw', 'roll', 'x', 'y', 'z', 'Score'])
        test_prediction = test_prediction.fillna('')

        valid_df = test_prediction

        train_df = pd.read_csv(self.path + 'train.csv')
        train_df = train_df[train_df.ImageId.isin(test_prediction.ImageId.unique())]
        # data description page says, The pose information is formatted as
        # model type, yaw, pitch, roll, x, y, z
        # but it doesn't, and it should be
        # model type, pitch, yaw, roll, x, y, z
        expanded_train_df = expand_df(train_df, ['model_type', 'pitch', 'yaw', 'roll', 'x', 'y', 'z'])

        max_workers = 10
        n_gt = len(expanded_train_df)
        ap_list = []
        # p = Pool(processes=max_workers)
        for result_flg, scores in (check_match(i) for i in range(10)):
            n_tp = np.sum(result_flg)
            recall = n_tp / n_gt
            ap = average_precision_score(result_flg, scores) * recall
            ap_list.append(ap)
        map = np.mean(ap_list)
        print(colored('mAP score is :', 'blue'), map)

    @staticmethod
    def show_image(image):
        # plt.ion()
        plt.figure()
        plt.title('processed image')
        img = image
        if len(img.shape) == 3:
            if img.shape[0] == 3:
                img = np.rollaxis(img, 0, 3)
            else:
                img = img[0, :, :]

        plt.imshow(img)
        plt.show()
        # plt.pause(0.01)
        # plt.clf()

    def make_log_dir(self):
        now = datetime.now()
        dt_string = now.strftime("%Y_%m_%d_%H_%M_%S")
        repository_name = self._settings.log_path + dt_string + '_' + str(self._settings.batch_size)
        os.makedirs(repository_name)
        return repository_name



Project/src/DataProcessing.py
import numpy as np
import pandas as pd
import math
import cv2
from sklearn.linear_model import LinearRegression
from skimage.filters import gaussian
from skimage.exposure import rescale_intensity
from skimage.util import random_noise
from torch.utils.data import Dataset
from scipy.optimize import minimize
from math import sin, cos
import matplotlib.pyplot as plt
import torch
from termcolor import colored


class DataProcessing(Dataset):
    def __init__(self, settings, dataframe, root_dir, data_agument=None, training=False):
        self._settings = settings
        self.df = dataframe
        self.root_dir = root_dir
        self.data_agument = data_agument
        self.train_dataset = None

        self.training = training
        self.path = self._settings.path
        self.camera_matrix = np.array([[2304.5479, 0, 1686.2379],
                                       [0, 2305.8757, 1354.9849],
                                       [0, 0, 1]], dtype=np.float32)
        self.img_width = self._settings.img_width
        self.img_height = self.img_width // 16 * 5
        self.model_scale = self._settings.model_scale

        self.xzy_slope =None

    def __len__(self):
        return len(self.df)

    def __getitem__(self, idx):
        if torch.is_tensor(idx):
            idx = idx.tolist()

        # Get image name
        idx, labels = self.df.values[idx]
        img_name = self.root_dir.format(idx)
        mask_name = img_name.replace('images', 'masks')

        # Read image
        img0 = self.imread(img_name, is_color=True, fast_mode=True)
        mask = self.imread(mask_name, is_color=False, fast_mode=True)
        try:
            imagemaskinv = cv2.bitwise_not(mask)
            img = cv2.bitwise_and(img0, img0, mask=imagemaskinv)
        except:
            img = img0
        img, flip = self.preprocess_image(img, data_agument=self.data_agument)

        img = np.rollaxis(img, 2, 0)

        # Get heatmap and regression maps
        heatmap, regr = self.get_mask_and_regr(img0, labels, flip=flip)
        regr = np.rollaxis(regr, 2, 0)

        return [img, heatmap, regr]

    def imread(self, path, is_color=True, fast_mode=False):
        img = cv2.imread(path, is_color)
        if not fast_mode and img is not None and len(img.shape) == 3:
            img = np.array(img[:, :, ::-1])  # inverse load image
        return img

    def preprocess_image(self, img, data_agument=[]):
        img = img[img.shape[0] // 2:]
        bg = np.ones_like(img) * img.mean(1, keepdims=True).astype(img.dtype)
        bg = bg[:, :img.shape[1] // 6]
        img = np.concatenate([bg, img, bg], 1)
        img = cv2.resize(img, (self.img_width, self.img_height))
        flip = False
        if self.training:
            np.random.seed(1)
            ratio = 1 // self._settings.data_agument_ratio
            if np.random.randint(ratio) == 0 and 'flip' in data_agument:
                img = img[:, ::-1]
                flip = True
            if np.random.randint(ratio * 2) == 0 and 'noise' in data_agument:
                img = random_noise(img, mode='gaussian', var=0.005)  # var default is 0.01
            if np.random.randint(ratio * 2) == 0 and 'blur' in data_agument:
                img = rescale_intensity(gaussian(img, sigma=1, multichannel=True))

        mean = [0.485, 0.456, 0.406]
        std = [0.229, 0.224, 0.225]
        img = (img / 255).astype('float32')
        for i in range(3):
            m = np.mean(img[:, :, i])
            s = np.std(img[:, :, i])
            img[:, :, i] = ((img[:, :, i] - m) * std[i] / s + mean[i]).astype('float32')
        return img, flip

    def get_mask_and_regr(self, img, labels, flip=False):
        """
        Get classification mask and regression values
        :param img: original image
        :param labels: prediction string
        :param flip: if flip image or not
        :return:
        """
        sigma = 0.8
        mask = np.zeros([self.img_height // self.model_scale, self.img_width // self.model_scale], dtype='float32')
        regr_names = ['x', 'y', 'z', 'yaw', 'pitch', 'roll']
        regr = np.zeros([self.img_height // self.model_scale, self.img_width // self.model_scale, 7], dtype='float32')
        coords = self.str2coords(labels)
        xs, ys = self.get_img_coords(labels)
        heatmap = np.zeros([self.img_height // self.model_scale, self.img_width // self.model_scale], dtype='float32')

        for x, y, regr_dict in zip(xs, ys, coords):
            x, y = y, x
            x = (x - img.shape[0] // 2) * self.img_height / (img.shape[0] // 2) / self.model_scale
            x = np.round(x).astype('int')
            y = (y + img.shape[1] // 6) * self.img_width / (img.shape[1] * 4 / 3) / self.model_scale
            y = np.round(y).astype('int')
            X1 = np.linspace(1, self.img_width // self.model_scale, self.img_width // self.model_scale)
            Y1 = np.linspace(1, self.img_height // self.model_scale, self.img_height // self.model_scale)
            [X, Y] = np.meshgrid(X1, Y1)
            X = X - math.floor(y) - 1
            Y = Y - math.floor(x) - 1
            D2 = X * X + Y * Y
            E2 = 2 * sigma ** 2
            Exponent = D2 / E2
            heatmap_ = np.exp(-Exponent)
            heatmap_ = heatmap_[:, :]
            heatmap[:, :] = np.maximum(heatmap[:, :], heatmap_[:, :])
            if x >= 0 and x < self.img_height // self.model_scale and y >= 0 and y < self.img_width // self.model_scale:
                regr_dict = self._regr_preprocess(regr_dict, flip)
                regr[x, y] = [regr_dict[n] for n in sorted(regr_dict)]
                mask[x, y] = 1
        if flip:
            heatmap = np.array(heatmap[:, ::-1])
            regr = np.array(regr[:, ::-1])
            mask = np.array(mask[:, ::-1])
        return heatmap, regr

    @staticmethod
    def str2coords(s, names=['id', 'yaw', 'pitch', 'roll', 'x', 'y', 'z']):
        '''
        Input:
            s: PredictionString (e.g. from train dataframe)
            names: array of what to extract from the string
        Output:
            list of dicts with keys from `names` : [['id':, 'yaw':, ...], [],...[]]
        '''
        coords = []
        for l in np.array(s.split()).reshape([-1, 7]):
            coords.append(dict(zip(names, l.astype('float'))))
            if 'id' in coords[-1]:
                coords[-1]['id'] = int(coords[-1]['id'])  # change id to int dtype
        return coords

    @staticmethod
    def coords2str(coords, names=['yaw', 'pitch', 'roll', 'x', 'y', 'z', 'confidence']):
        s = []
        for c in coords:
            for n in names:
                s.append(str(c.get(n, 0)))
        return ' '.join(s)

    def get_img_coords(self, s):
        """
        Convert world coordinates to pixel coordinates
        :param s: Input is a PredictionString (e.g. from train dataframe)
        :return: Output is two arrays:
            xs: x coordinates in the image
            ys: y coordinates in the image
        """
        coords = self.str2coords(s)
        xs = [c['x'] for c in coords]
        ys = [c['y'] for c in coords]
        zs = [c['z'] for c in coords]
        P = np.array(list(zip(xs, ys, zs))).T  # row: x, y, z
        img_p = np.dot(self.camera_matrix, P).T
        img_p[:, 0] /= img_p[:, 2]
        img_p[:, 1] /= img_p[:, 2]
        img_xs = img_p[:, 0]
        img_ys = img_p[:, 1]
        img_zs = img_p[:, 2]  # z = Distance from the camera
        return img_xs, img_ys

    def _regr_preprocess(self, regr_dict, flip=False):
        """

        :param regr_dict:
        :param flip:
        :return: regr_dict.keys() = roll, pitch_sin, pitch_cos, x, y, z
        """
        if flip:
            for k in ['x', 'pitch', 'roll']:
                regr_dict[k] = -regr_dict[k]
        for name in ['x', 'y', 'z']:
            regr_dict[name] = regr_dict[name] / 100
        regr_dict['roll'] = self.rotate(regr_dict['roll'], np.pi)
        regr_dict['pitch_sin'] = sin(regr_dict['pitch'])
        regr_dict['pitch_cos'] = cos(regr_dict['pitch'])
        regr_dict.pop('pitch')
        regr_dict.pop('id')
        return regr_dict

    def rotate(self, x, angle):  # rotate angle
        x = x + angle
        x = x - (x + np.pi) // (2 * np.pi) * 2 * np.pi
        return x

    # convert euler angle to rotation matrix
    def euler_to_Rot(self, yaw, pitch, roll):
        Y = np.array([[cos(yaw), 0, sin(yaw)],
                      [0, 1, 0],
                      [-sin(yaw), 0, cos(yaw)]])
        P = np.array([[1, 0, 0],
                      [0, cos(pitch), -sin(pitch)],
                      [0, sin(pitch), cos(pitch)]])
        R = np.array([[cos(roll), -sin(roll), 0],
                      [sin(roll), cos(roll), 0],
                      [0, 0, 1]])
        return np.dot(Y, np.dot(P, R))

    def draw_line(self, image, points):
        color = (255, 0, 0)
        cv2.line(image, tuple(points[0][:2]), tuple(points[3][:2]), color, 16)
        cv2.line(image, tuple(points[0][:2]), tuple(points[1][:2]), color, 16)
        cv2.line(image, tuple(points[1][:2]), tuple(points[2][:2]), color, 16)
        cv2.line(image, tuple(points[2][:2]), tuple(points[3][:2]), color, 16)
        return image

    def draw_points(self, image, points):
        for (p_x, p_y, p_z) in points:
            cv2.circle(image, (p_x, p_y), int(1000 / p_z), (0, 255, 0), -1)  # the car is far -> p_z large -> radius is small
        #         if p_x > image.shape[1] or p_y > image.shape[0]:
        #             print('Point', p_x, p_y, 'is out of image with shape', image.shape)
        return image

    def visualize(self, img, coords):
        x_l = 1.02
        y_l = 0.80
        z_l = 2.31
        img = img.copy()
        for point in coords:
            # Get values
            x, y, z = point['x'], point['y'], point['z']
            yaw, pitch, roll = -point['pitch'], -point['yaw'], -point['roll']
            # Math
            Rt = np.eye(4)  # 4*4对角矩阵
            t = np.array([x, y, z])
            Rt[:3, 3] = t
            Rt[:3, :3] = self.euler_to_Rot(yaw, pitch, roll).T
            Rt = Rt[:3, :]
            P = np.array([[x_l, -y_l, -z_l, 1],
                          [x_l, -y_l, z_l, 1],
                          [-x_l, -y_l, z_l, 1],
                          [-x_l, -y_l, -z_l, 1],
                          [0, 0, 0, 1]]).T
            img_cor_points = np.dot(self.camera_matrix, np.dot(Rt, P))
            img_cor_points = img_cor_points.T
            img_cor_points[:, 0] /= img_cor_points[:, 2]
            img_cor_points[:, 1] /= img_cor_points[:, 2]
            img_cor_points = img_cor_points.astype(int)
            # Drawing
            img = self.draw_line(img, img_cor_points)
            img = self.draw_points(img, img_cor_points[-1:])

        return img

    def _regr_back(self, regr_dict):
        for name in ['x', 'y', 'z']:
            regr_dict[name] = regr_dict[name] * 100
        regr_dict['roll'] = self.rotate(regr_dict['roll'], -np.pi)

        pitch_sin = regr_dict['pitch_sin'] / np.sqrt(regr_dict['pitch_sin'] ** 2 + regr_dict['pitch_cos'] ** 2)
        pitch_cos = regr_dict['pitch_cos'] / np.sqrt(regr_dict['pitch_sin'] ** 2 + regr_dict['pitch_cos'] ** 2)
        regr_dict['pitch'] = np.arccos(pitch_cos) * np.sign(pitch_sin)
        return regr_dict


    def convert_3d_to_2d(self, x, y, z, fx=2304.5479, fy=2305.8757, cx=1686.2379, cy=1354.9849):
        # stolen from https://www.kaggle.com/theshockwaverider/eda-visualization-baseline
        return x * fx / z + cx, y * fy / z + cy

    def optimize_xy(self, r, c, x0, y0, z0, flipped=False):
        def distance_fn(xyz):
            img = self.imread(self.path + 'train_images/ID_8a6e65317' + '.jpg')
            IMG_SHAPE = img.shape
            x, y, z = xyz
            xx = -x if flipped else x
            slope_err = (self.xzy_slope.predict([[xx, z]])[0] - y) ** 2
            x, y = self.convert_3d_to_2d(x, y, z0)
            y, x = x, y
            x = (x - IMG_SHAPE[0] // 2) * self.img_height / (IMG_SHAPE[0] // 2) / self.model_scale
            # x = np.round(x).astype('int')
            y = (y + IMG_SHAPE[1] // 6) * self.img_width / (IMG_SHAPE[1] * 4 / 3) / self.model_scale
            # y = y * self.img_width / IMG_SHAPE[1] / self.model_scale
            # y = np.round(y).astype('int')
            # return (x - r) ** 2 + (y - c) ** 2
            return max(0.2, (x - r) ** 2 + (y - c) ** 2) + max(0.4, slope_err)
        # print(colored('Finding Minimize.....', 'red'))
        # res = minimize(distance_fn, [x0, y0, z0], method='Powell')
        res = minimize(distance_fn, [x0, y0, z0], method='Powell')
        # print(colored('Minimization Done!', 'red'))
        x_new, y_new, z_new = res.x
        return x_new, y_new, z_new

    def clear_duplicates(self, coords):
        for c1 in coords:
            xyz1 = np.array([c1['x'], c1['y'], c1['z']])
            for c2 in coords:
                xyz2 = np.array([c2['x'], c2['y'], c2['z']])
                distance = np.sqrt(((xyz1 - xyz2) ** 2).sum())
                if distance < self._settings.distance_thresh_clear:
                    if c1['confidence'] < c2['confidence']:
                        c1['confidence'] = -1
        return [c for c in coords if c['confidence'] > 0]

    def extract_coords(self, prediction, flipped=False):
        logits = prediction[0]
        regr_output = prediction[1:]
        points = np.argwhere(logits > 0.5)
        col_names = sorted(['x', 'y', 'z', 'yaw', 'pitch_sin', 'pitch_cos', 'roll'])
        coords = []
        for r, c in points:  # r, c is coordinate where logits is larger than 0
            regr_dict = dict(zip(col_names, regr_output[:, r, c]))
            coords.append(self._regr_back(regr_dict))
            coords[-1]['confidence'] = 1 / (1 + np.exp(-logits[r, c]))
            if self._settings.optimize_coordinate:
                coords[-1]['x'], coords[-1]['y'], coords[-1]['z'] = self.optimize_xy(r, c, coords[-1]['x'], coords[-1]['y'],
                                                                                coords[-1]['z'], flipped)
        coords = self.clear_duplicates(coords)
        return coords

    def point_regr(self):
        points_df = pd.DataFrame()
        for col in ['x', 'y', 'z', 'yaw', 'pitch', 'roll']:
            arr = []
            for ps in self.train_dataset['PredictionString']:
                coords = self.str2coords(ps)
                arr += [c[col] for c in coords]
            points_df[col] = arr
        xzy_slope = LinearRegression()
        X = points_df[['x', 'z']]
        y = points_df['y']
        print(colored('Fitting Linear Regression.......', 'red'))
        xzy_slope.fit(X, y)
        print(colored('Finish Fitting Linear Regression!', 'red'))
        # print('MAE with x:', mean_absolute_error(y, xzy_slope.predict(X)))
        # print('\ndy/dx = {:.3f}\ndy/dz = {:.3f}'.format(*xzy_slope.coef_))
        self.xzy_slope = xzy_slope

    def show_result(self, idx, coords_true, coords_pred, df_dev, dir_name):
        train_images_dir = self.path + 'train_images/{}.jpg'
        img = self.imread(train_images_dir.format(df_dev['ImageId'].iloc[idx]))
        fig, axes = plt.subplots(1, 2, figsize=(30, 30))
        axes[0].set_title('Ground truth')
        axes[0].imshow(self.visualize(img, coords_true))
        axes[1].set_title('Prediction')
        axes[1].imshow(self.visualize(img, coords_pred))
        plt.savefig(dir_name + '/' + str(idx) + '_result.png')